### DISCUSSION TITLE
Multiple Singleton instances

### DISCUSSION THREAD

**Discussion by omgMath** [vercel/next.js#31496] [Help]
Hi

We've a `next.js` app running. To avoid unnecessary requests to the CMS, we've implemented a cache, which we clear after changes in the CMS by calling a route `api/clear-cache`.
Simplified version: This route basically takes a exported, already instantiated class `Cache` and clears it.

On the other side, when sending a request to the CMS, we await the response and store that in the same exported, already instantiated class `Cache`.

For debugging purposes, I've added a UUID to the `Cache` instance, and it shows:
- on  `api/clear-cache`: Clear cache with id: b5583b47-9eef-4cd1-a0ad-b24508ee1e71
- on `getStaticProps` (with `revalidate: 20`): Get from cache with id: fcbe712a-e4ab-4789-a3a8-4b925754a329

I don't really understand what is happening and hoped someone of you could make sense of it...? Are there some child processes spawning? Does someone see how to solve it?

Thanks!

**Comment by SLTKA** [comment]
We have hit the same issue and found that reason for it that bundle for API routes would be separate to bundle for pages, so we have 2 instances of the same code and classic JS singleton approach doesn't work. We are exploring usage of `global` space but need to do more testing if it is safe to use.

It is couple years passed since your post @omgMath , do you have answer to it by any chance?

**Reply by omgMath** [reply]
Hi @SLTKA
We were not able to find a solution to the actual problem. But in the meantime [on demand validation](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#using-on-demand-revalidation) was introduced, which made our use for a cache obsolete. 
Hope you find a solution for your use case as well!

**Reply by SLTKA** [reply]
We have solved it using `globalThis` like:

```

import { MyClass } from './MyClass';

/** singleton */
globalThis.__sharedClassInstance = globalThis.__sharedClassInstance ?? new MyClass(...);

// Expose for consumption
export const myClassSIngletone = globalThis.__sharedClassInstance;
```

```
//global types in any *.d.ts in the project

declare var __sharedClassInstance:
  | undefined
  | import('./lib/MyClass').MyClass;
```

This works for all cases except when NextJS prebuilds pages as part of SSG using [Jest Workers](https://www.npmjs.com/package/jest-worker). They have isolated memory and would not reuse anything. Curious if there is a way to pass additional data to them from the main process...
